symfony/plugins/orangehrmTimePlugin/Dao/CustomerDao.php

```diff
@@ -184,4 +207,101 @@ public function getCustomerIdListForProjectAdmin(int $projectAdminEmpNumber, boo
         $result = $q->getQuery()->getArrayResult();
         return array_column($result, 'id');
     }
+
+    /**
+     * @param int $customerId
+     * @return bool
+     */
+    public function hasCustomerGotTimesheetItems(int $customerId): bool
+    {
+        $q = $this->createQueryBuilder(TimesheetItem::class, 'timesheetItem');
+        $q->leftJoin('timesheetItem.project', 'project');
+        $q->leftJoin('project.customer', 'customer');
+        $q->andWhere('customer.id = :customerId');
+        $q->setParameter('customerId', $customerId);
+        $count = $this->getPaginator($q)->count();
+        return ($count > 0);
+    }
+
+    /**
+     * @return int[]
+     */
+    public function getUnselectableCustomerIds(): array
+    {
+        $q = $this->createQueryBuilder(TimesheetItem::class, 'timesheetItem');
+        $q->leftJoin('timesheetItem.project', 'project');
+        $q->leftJoin('project.customer', 'customer');
+        $q->select('customer.id');
+        $q->groupBy('customer.id');
+        $result = $q->getQuery()->getArrayResult();
+        return array_column($result, 'id');
+    }
+
+    /**
+     * @param int[] $customerIds
+     * @return void
+     */
+    public function deleteRelativeProjectsForCustomer(array $customerIds): void
+    {
+        $q = $this->createQueryBuilder(Project::class, 'project');
+        $q->andWhere($q->expr()->in('project.customer', ':customerIds'));
+        $q->setParameter('customerIds', $customerIds);
+        $q->andWhere('project.deleted = :deleted');
+        $q->setParameter('deleted', false);
+
+        /* @var Project[] $projects */
+        $projects = $q->getQuery()->execute();
+        $projectIds = [];
+        foreach ($projects as $project) {
+            $projectIds[] = $project->getId();
+        }
+        // Delete the records that belong to the customer
+        $this->deleteProjects($projectIds);
+        $this->deleteRelativeProjectActivitiesForProject($projectIds);
+        $this->deleteRelativeProjectAdminsForProject($projectIds);
```

Comment:
don't we need to wrap inside transaction

Reply:
https://github.com/orangehrm/orangehrm/pull/1147/files#diff-2246fcd120f04977a00e3effae04574453c35d7bd240aa87ed07094f8b4ce4bbR55-R78 ???